src/System.Runtime.Numerics/src/System/Numerics/Complex.cs

```diff
@@ -110,31 +113,46 @@ public static Complex Divide(Complex dividend, Complex divisor)
 
         public static double Abs(Complex value)
         {
-            if (double.IsInfinity(value._real) || double.IsInfinity(value._imaginary))
+            return Hypot(value._real, value._imaginary);
+        }
+
+        private static double Hypot(double a, double b)
+        {
+            // Using
+            //   sqrt(a^2 + b^2) = |a| * sqrt(a^2/a^2 + b^2/a^2)
+            // we can factor out the larger component to dodge overflow even when a * a would overflow.
+
+            a = Math.Abs(a);
+            b = Math.Abs(b);
+
+            double small, large;
+            if (a < b)
             {
-                return double.PositiveInfinity;
+                small = a;
+                large = b;
+            }
+            else
+            {
+                small = b;
+                large = a;
             }
 
-            // |value| == sqrt(a^2 + b^2)
-            // sqrt(a^2 + b^2) == a/a * sqrt(a^2 + b^2) = a * sqrt(a^2/a^2 + b^2/a^2)
-            // Using the above we can factor out the square of the larger component to dodge overflow.
-            double c = Math.Abs(value._real);
-            double d = Math.Abs(value._imaginary);
-
-            if (c > d)
+            if (small == 0.0)
             {
-                double r = d / c;
-                return c * Math.Sqrt(1.0 + r * r);
+                return (large);
             }
-            else if (d == 0.0)
+            else if (Double.IsPositiveInfinity(large) && !Double.IsNaN(small))
             {
-                return c;  // c is either 0.0 or NaN
+                // The NaN test is necessary so we don't return +inf when small=NaN and large=+inf.
+                // NaN in any other place returns NaN without any special handling.
```

Comment:
Are there any other operations on `Complex` which have made this decision about whether or not to propogate NaNs? If so, we should follow that pattern. If not, we should try to follow what other libraries do (although based on my testing below, that is not at all consistent...)